File: typescript/src/core/puzzle2d.ts
import { Piece } from './piece';

export class Puzzle2D {
  private readonly PIECES = 12;
  private ROWS = 0;
  private COLUMNS = 0;
  private totalSolutions = 0;
  private triedPieces = 0;
  private row = 0;
  private column = 0;
  private piecesIndices: number[] = [];
  private solution: number[] = new Array(this.PIECES);
  private pieces: Piece[] = new Array(this.PIECES);
  private names = '';
  private grid: number[][] = [];
  private currPiece: Piece | null = null;
  private totalFillInGrid = 0;
  private availInGrid = 0;
  private aborted = false;

  set(rows: number, columns: number): void {
    Piece.totalFill = 0;
    this.piecesIndices = [];
    this.solution = new Array(this.PIECES);
    this.totalSolutions = 0;
    this.triedPieces = 0;
    this.aborted = false;

    this.ROWS = rows;
    this.COLUMNS = columns;

    this.grid = new Array(this.ROWS);
    for (let i = 0; i < this.ROWS; i++) {
      this.grid[i] = new Array(this.COLUMNS).fill(0);
    }

    const allPieces: number[][][] = [
      [[1], [1], [1], [1, 1]],
      [[1, 1], [1], [1, 1]],
      [[0, 1, 1], [1, 1], [0, 1]],
      [[0, 1], [1, 1, 1], [0, 1]],
      [[1, 1, 1, 1], [0, 0, 1]],
      [[0, 1], [1, 1], [1], [1]],
      [[0, 0, 1], [0, 1, 1], [1, 1]],
      [[1], [1, 1], [1, 1]],
      [[0, 0, 1], [1, 1, 1], [1]],
      [[0, 0, 1], [0, 0, 1], [1, 1, 1]],
      [[0, 0, 1], [1, 1, 1], [0, 0, 1]],
      [[1, 1, 1, 1, 1]],
    ];

    const rotations = [4, 4, 2, 1, 4, 4, 4, 4, 2, 4, 4, 2];
    const symmetric = [2, 1, 1, 1, 2, 2, 1, 2, 2, 1, 1, 1];
    this.names = 'LUFXYNWPZVTI';

    if (this.ROWS === this.COLUMNS && this.ROWS === 8) {
      rotations[2] = 1;
    }

    if (this.ROWS === 0 && this.COLUMNS === 0) {
      throw new Error('not supported yet');
    } else {
      for (let i = 0; i < this.PIECES; i++) {
        this.piecesIndices.push(i);
        this.pieces[i] = new Piece(i, allPieces[i], rotations[i], symmetric[i], this.names[i]);
      }
      if (this.ROWS === this.COLUMNS && this.ROWS === 8) {
        this.grid[0][0] = -1;
        this.grid[7][0] = -1;
        this.grid[0][7] = -1;
        this.grid[7][7] = -1;
      }
      while (this.grid[0][this.column] === -1) {
        this.column++;
      }
    }

    this.totalFillInGrid = this.ROWS * this.COLUMNS;
    for (const row of this.grid) {
      for (const val of row) {
        if (val === -1) {
          this.totalFillInGrid--;
        }
      }
    }

    this.availInGrid = this.totalFillInGrid;
    console.log(`Found ${this.ROWS} rows, ${this.COLUMNS} cols, with total of cells ${this.availInGrid}`);
    this.showGrid();
  }

  private showGrid(): void {
    console.log(`[msec] showGrid. Tried Pieces ${this.triedPieces} leftPieces ${this.piecesIndices.length}`);
    for (let i = 0; i < this.ROWS; i++) {
      let line = '';
      for (let j = 0; j < this.COLUMNS; j++) {
        if (this.grid[i][j] === -1) {
          line += '*  ';
        } else if (this.grid[i][j] === 0) {
          line += '-  ';
        } else {
          line += `${this.names[this.grid[i][j] - 1]} `;
        }
      }
      console.log(line);
    }
  }

  private showPieces(): void {
    let line = '';
    for (let i = 0; i < this.PIECES - this.piecesIndices.length; i++) {
      line += `${this.names[this.solution[i]]} `;
    }
    console.log(line);
  }

  private put(): void {
    const leftPieces = this.piecesIndices.length;

    if (leftPieces === 0) {
      this.totalSolutions++;
      console.log(`totalSolutions ${this.totalSolutions}`);
      if (this.totalSolutions === 1) {
        console.log('Found a solution');
        this.showGrid();
        this.showPieces();
      }
    }

    if (this.totalSolutions >= 1) return;

    const rowsSet = new Array(5);
    const columnsSet = new Array(5);

    for (let i = 0; i < leftPieces; i++) {
      const pieceIdx = this.piecesIndices[i];
      this.currPiece = this.pieces[pieceIdx];
      for (let r = this.currPiece.getAvailRotations(); r >= 1; r--) {
        if (this.canPut(rowsSet, columnsSet)) {
          this.piecesIndices.splice(i, 1);
          this.solution[this.PIECES - leftPieces] = pieceIdx;
          this.putCurrPiece(rowsSet, columnsSet);

          if (this.triedPieces % 50000 === 0) {
            this.showGrid();
            this.showPieces();

            if (this.aborted) {
              console.log(`Id ${this.ROWS}_${this.COLUMNS}. Signaled timed out! totalSolutions ${this.totalSolutions}`);
              return;
            }
          }

          this.put();
          this.removeLast(pieceIdx, rowsSet, columnsSet);
          this.piecesIndices.splice(i, 0, pieceIdx);
        }
        this.currPiece.rotate();
      }
    }
  }

  private putCurrPiece(rowsSet: number[], columnsSet: number[]): void {
    this.currPiece!.setPosition(this.row, this.column);

    const currIndex = this.currPiece!.index + 1;
    for (let i = 0; i < this.currPiece!.totalThisFill; i++) {
      this.grid[rowsSet[i]][columnsSet[i]] = currIndex;
    }

    this.goForward();
    this.availInGrid -= this.currPiece!.totalThisFill;
  }

  private goForward(): void {
    while (this.row < this.ROWS) {
      while (this.column < this.COLUMNS) {
        if (this.grid[this.row][this.column] === 0) return;
        this.column++;
      }
      this.column = 0;
      this.row++;
    }
  }

  private removeLast(piece: number, rowsSet: number[], columnsSet: number[]): void {
    this.currPiece = this.pieces[piece];

    this.row = this.currPiece.getRow();
    this.column = this.currPiece.getColumn();

    for (let i = 0; i < this.currPiece.totalThisFill; i++) {
      this.grid[rowsSet[i]][columnsSet[i]] = 0;
    }

    this.availInGrid += this.currPiece.totalThisFill;
  }

  private canPut(rowsSet: number[], columnsSet: number[]): boolean {
    try {
      let setSoFar = 0;
      const j = this.currPiece!.getFirstSquarePos();
      const columnjj = this.column - j;
      for (let i = 0; i < this.currPiece!.totalThisFill; i++) {
        const rowi = this.row + this.currPiece!.getRowSet(setSoFar);
        const columnj = columnjj + this.currPiece!.getColumnSet(setSoFar);
        if (this.grid[rowi][columnj] !== 0) {
          return false;
        } else {
          rowsSet[setSoFar] = rowi;
          columnsSet[setSoFar] = columnj;
          setSoFar++;
        }
      }
    } catch (_) {
      return false;
    }

    this.triedPieces++;
    return true;
  }

  solve(signal?: AbortSignal): number {
    const start = Date.now();

    if (signal) {
      signal.addEventListener('abort', () => {
        this.aborted = true;
      });
    }

    if (this.totalFillInGrid !== Piece.totalFill) {
      console.error(`Invalid config, grid ${this.totalFillInGrid} pieces ${Piece.totalFill}`);
    } else {
      Piece.totalFill = 0;
      console.log(`Starting rows ${this.ROWS} cols ${this.COLUMNS}`);
      this.put();
    }

    const elapsedTime = Math.floor((Date.now() - start) / 1000);
    console.log(`tried ${this.triedPieces.toLocaleString()} pieces`);
    if (elapsedTime > 0) {
      console.log(`at ${(this.triedPieces / elapsedTime).toLocaleString()} pieces per sec`);
    }
    console.log(`number of solutions ${this.totalSolutions * (this.ROWS === this.COLUMNS ? 8 : 4)}`);

    return this.totalSolutions;
  }
}

File: go/core/puzzle2d.go
package core

import (
	"fmt"
	"sync/atomic"
	"time"
)

type Puzzle2D struct {
	PIECES          int
	ROWS            int
	COLUMNS         int
	TotalSolutions  int
	TriedPieces     int
	Row             int
	Column          int
	PiecesIndices   []int
	Solution        []int
	Pieces          []*Piece
	Names           string
	Grid            [][]int
	CurrPiece       *Piece
	TotalFillInGrid int
	AvailInGrid     int
	aborted         *atomic.Bool
}

func NewPuzzle2D() *Puzzle2D {
	return &Puzzle2D{
		PIECES:  12,
		aborted: &atomic.Bool{},
	}
}

func (p *Puzzle2D) Set(rows, columns int) {
	TotalFill = 0
	p.PiecesIndices = []int{}
	p.Solution = make([]int, p.PIECES)
	p.TotalSolutions = 0
	p.TriedPieces = 0
	p.aborted.Store(false)

	p.ROWS = rows
	p.COLUMNS = columns

	p.Grid = make([][]int, p.ROWS)
	for i := range p.Grid {
		p.Grid[i] = make([]int, p.COLUMNS)
	}

	allPieces := [][][]int{
		{{1}, {1}, {1}, {1, 1}},
		{{1, 1}, {1}, {1, 1}},
		{{0, 1, 1}, {1, 1}, {0, 1}},
		{{0, 1}, {1, 1, 1}, {0, 1}},
		{{1, 1, 1, 1}, {0, 0, 1}},
		{{0, 1}, {1, 1}, {1}, {1}},
		{{0, 0, 1}, {0, 1, 1}, {1, 1}},
		{{1}, {1, 1}, {1, 1}},
		{{0, 0, 1}, {1, 1, 1}, {1}},
		{{0, 0, 1}, {0, 0, 1}, {1, 1, 1}},
		{{0, 0, 1}, {1, 1, 1}, {0, 0, 1}},
		{{1, 1, 1, 1, 1}},
	}

	rotations := []int{4, 4, 2, 1, 4, 4, 4, 4, 2, 4, 4, 2}
	symmetric := []int{2, 1, 1, 1, 2, 2, 1, 2, 2, 1, 1, 1}
	p.Names = "LUFXYNWPZVTI"

	if p.ROWS == p.COLUMNS && p.ROWS == 8 {
		rotations[2] = 1
	}

	if p.ROWS == 0 && p.COLUMNS == 0 {
		panic("not supported yet")
	}

	p.Pieces = make([]*Piece, p.PIECES)
	for i := 0; i < p.PIECES; i++ {
		p.PiecesIndices = append(p.PiecesIndices, i)
		p.Pieces[i] = NewPiece(i, allPieces[i], rotations[i], symmetric[i], string(p.Names[i]))
	}

	if p.ROWS == p.COLUMNS && p.ROWS == 8 {
		p.Grid[0][0] = -1
		p.Grid[7][0] = -1
		p.Grid[0][7] = -1
		p.Grid[7][7] = -1
	}
	for p.Grid[0][p.Column] == -1 {
		p.Column++
	}

	p.TotalFillInGrid = p.ROWS * p.COLUMNS
	for _, row := range p.Grid {
		for _, val := range row {
			if val == -1 {
				p.TotalFillInGrid--
			}
		}
	}

	p.AvailInGrid = p.TotalFillInGrid
	fmt.Printf("Found %d rows, %d cols, with total of cells %d\n", p.ROWS, p.COLUMNS, p.AvailInGrid)
	p.ShowGrid()
}

func (p *Puzzle2D) ShowGrid() {
	fmt.Printf("[msec] showGrid. Tried Pieces %d leftPieces %d\n", p.TriedPieces, len(p.PiecesIndices))
	for i := 0; i < p.ROWS; i++ {
		line := ""
		for j := 0; j < p.COLUMNS; j++ {
			if p.Grid[i][j] == -1 {
				line += "*  "
			} else if p.Grid[i][j] == 0 {
				line += "-  "
			} else {
				line += fmt.Sprintf("%c ", p.Names[p.Grid[i][j]-1])
			}
		}
		fmt.Println(line)
	}
}

func (p *Puzzle2D) showPieces() {
	line := ""
	for i := 0; i < p.PIECES-len(p.PiecesIndices); i++ {
		line += fmt.Sprintf("%c ", p.Names[p.Solution[i]])
	}
	fmt.Println(line)
}

func (p *Puzzle2D) Put() {
	leftPieces := len(p.PiecesIndices)

	if leftPieces == 0 {
		p.TotalSolutions++
		fmt.Printf("totalSolutions %d\n", p.TotalSolutions)
		if p.TotalSolutions == 1 {
			fmt.Println("Found a solution")
			p.ShowGrid()
			p.showPieces()
		}
	}

	if p.TotalSolutions >= 1 {
		return
	}

	rowsSet := make([]int, 5)
	columnsSet := make([]int, 5)

	for i := 0; i < leftPieces; i++ {
		pieceIdx := p.PiecesIndices[i]
		p.CurrPiece = p.Pieces[pieceIdx]
		for r := p.CurrPiece.GetAvailRotations(); r > 0; r-- {
			if p.CanPut(rowsSet, columnsSet) {
				p.PiecesIndices = append(p.PiecesIndices[:i], p.PiecesIndices[i+1:]...)
				p.Solution[p.PIECES-leftPieces] = pieceIdx
				p.PutCurrPiece(rowsSet, columnsSet)

				if p.TriedPieces%50000 == 0 {
					p.ShowGrid()
					p.showPieces()

					if p.aborted.Load() {
						fmt.Printf("Id %d_%d. Signaled timed out! totalSolutions %d\n", p.ROWS, p.COLUMNS, p.TotalSolutions)
						return
					}
				}

				p.Put()

				p.RemoveLast(pieceIdx, rowsSet, columnsSet)
				// insert back at position i
				newIndices := make([]int, 0, len(p.PiecesIndices)+1)
				newIndices = append(newIndices, p.PiecesIndices[:i]...)
				newIndices = append(newIndices, pieceIdx)
				newIndices = append(newIndices, p.PiecesIndices[i:]...)
				p.PiecesIndices = newIndices
			}
			p.CurrPiece.Rotate()
		}
	}
}

func (p *Puzzle2D) PutCurrPiece(rowsSet, columnsSet []int) {
	p.CurrPiece.SetPosition(p.Row, p.Column)

	currIndex := p.CurrPiece.Index + 1
	for i := 0; i < p.CurrPiece.TotalThisFill; i++ {
		p.Grid[rowsSet[i]][columnsSet[i]] = currIndex
	}

	p.GoForward()
	p.AvailInGrid -= p.CurrPiece.TotalThisFill
}

func (p *Puzzle2D) GoForward() {
	for p.Row < p.ROWS {
		for p.Column < p.COLUMNS {
			if p.Grid[p.Row][p.Column] == 0 {
				return
			}
			p.Column++
		}
		p.Column = 0
		p.Row++
	}
}

func (p *Puzzle2D) RemoveLast(piece int, rowsSet, columnsSet []int) {
	p.CurrPiece = p.Pieces[piece]

	p.Row = p.CurrPiece.GetRow()
	p.Column = p.CurrPiece.GetColumn()

	for i := 0; i < p.CurrPiece.TotalThisFill; i++ {
		p.Grid[rowsSet[i]][columnsSet[i]] = 0
	}

	p.AvailInGrid += p.CurrPiece.TotalThisFill
}

func (p *Puzzle2D) CanPut(rowsSet, columnsSet []int) bool {
	if p.CurrPiece == nil {
		return false
	}

	setSoFar := 0
	j := p.CurrPiece.GetFirstSquarePos()
	columnJj := p.Column - j

	for i := 0; i < p.CurrPiece.TotalThisFill; i++ {
		rowI := p.Row + p.CurrPiece.GetRowSet(setSoFar)
		columnJ := columnJj + p.CurrPiece.GetColumnSet(setSoFar)

		if rowI < 0 || rowI >= p.ROWS || columnJ < 0 || columnJ >= p.COLUMNS {
			return false
		}
		if p.Grid[rowI][columnJ] != 0 {
			return false
		}
		rowsSet[setSoFar] = rowI
		columnsSet[setSoFar] = columnJ
		setSoFar++
	}

	p.TriedPieces++
	return true
}

func (p *Puzzle2D) Solve(abortChan <-chan struct{}) int {
	start := time.Now()

	if abortChan != nil {
		select {
		case <-abortChan:
			p.aborted.Store(true)
		default:
		}
	}

	if p.TotalFillInGrid != TotalFill {
		fmt.Printf("Invalid config, grid %d pieces %d\n", p.TotalFillInGrid, TotalFill)
	} else {
		TotalFill = 0
		fmt.Printf("Starting rows %d cols %d\n", p.ROWS, p.COLUMNS)
		p.Put()
	}

	elapsedTime := int(time.Since(start).Seconds())
	fmt.Printf("tried %d pieces\n", p.TriedPieces)
	if elapsedTime > 0 {
		fmt.Printf("at %d pieces per sec\n", p.TriedPieces/elapsedTime)
	}
	multiplier := 4
	if p.ROWS == p.COLUMNS {
		multiplier = 8
	}
	fmt.Printf("number of solutions %d\n", p.TotalSolutions*multiplier)

	return p.TotalSolutions
}
